src/sequence_manager.cpp

```diff
@@ -44,14 +46,35 @@ bool SequenceManager::sequenceExists(const uint64_t sequenceId) const {
     return sequences.count(sequenceId);
 }
 
-Status SequenceManager::removeTimedOutSequences(std::chrono::steady_clock::time_point currentTime) {
+Status SequenceManager::removeTimedOutSequences() {
+    std::unique_lock<std::mutex> sequenceManagerLock(mutex);
     for (auto it = sequences.cbegin(); it != sequences.cend();) {
-        auto& sequence = it->second;
-        auto timeDiff = currentTime - sequence.getLastActivityTime();
-        if (std::chrono::duration_cast<std::chrono::seconds>(timeDiff).count() > timeout)
+        Sequence& sequence = getSequence(it->second.getId());
+        if (sequence.isTimedOut()) {
+            SPDLOG_LOGGER_DEBUG(sequence_manager_logger, "Sequence watcher thread for model {} Removing timeouted sequence - Id: {}", modelName, sequence.getId());
             it = sequences.erase(it);
-        else
+        } else {
             ++it;
+        }
+    }
+    return StatusCode::OK;
+}
+
+Status SequenceManager::checkForTimedOutSequences() {
+    std::unique_lock<std::mutex> sequenceManagerLock(mutex);
+    std::chrono::steady_clock::time_point currentTime = std::chrono::steady_clock::now();
+    for (auto it = sequences.cbegin(); it != sequences.cend();) {
+        Sequence& sequence = getSequence(it->second.getId());
+        std::unique_lock<std::mutex> sequenceLock(sequence.getMutex());
+        if (!sequence.isTerminated()) {
```

Comment:
Sequence gets set to terminated state in infer method during special inputs processing phase with sequence manager mutex locked. Since it doesn't get set to terminated state during request processing (within sequence critical section) I think we don't need to lock sequence before checking if it's terminated (we don't in infer() method). 
Can we move this condition up so it gets check before sequence locking attempt?